cluster-api/apis/cluster/types.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package cluster contains types to represent Kubernetes cluster configuration.
+package cluster // import "k8s.io/kube-deploy/cluster-api/apis/cluster"
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// Cluster is an API object representing a cluster's control-plane
+// configuration and status.
+type Cluster struct {
+	metav1.ObjectMeta
+	metav1.TypeMeta
+
+	Spec   ClusterSpec
+	Status ClusterStatus
+}
+
+type ClusterSpec struct {
+	// Nominal version of Kubernetes control-plane to run.
+	KubernetesVersion KubernetesVersionInfo
+
+	// Basic API configuration
+	APIConfig APIServerConfig
+
+	// Cluster network configuration
+	ClusterNetwork ClusterNetworkingConfig
+
+	// Provider-specific serialized configuration to use during
+	// cluster creation. It is recommended that providers maintain
+	// their own versioned API types that should be
+	// serialized/deserialized from this field.
+	ProviderConfig string
```

Comment:
I understand the usecases for this at the `Machine` level, but can you give a few examples of what different providers might want to include in the `ProviderConfig` in the `ClusterSpec`?

Reply:
Where is the Provider name specified?

Reply:
What about

```go
ProviderConfig interface{}
```

or

```go
type ProviderConfig interface{
    //stuff
}

// snip 

ProviderConfig ProviderConfig

```

Why is this a serialized string?

Reply:
Could something like `k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.Unstructured` work here? https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/unstructured/unstructured.go#L48

Reply:
@kris-nova @munnerz 
I'm not 100% sure, but I think that correct type should be either `runtime.RawExtension` or `runtime.Object`:
https://github.com/kubernetes/apimachinery/blob/18a564baac720819100827c16fdebcadb05b2d0d/pkg/runtime/types.go#L47-L100
`api-machinary` probably probably knows best.

Reply:
FYI, generators can generate DeepCopy method when `interface{} ` is used.